File: 2023/02/solution.pas
{ solution.pas -- cube colors  -- Troy Brumley BlameTroi@gmail.com }

program solution;

{ expects to be compiled with the following two mode flags and command
line/fpc.cfg options of -Ciort -Sgix -Sh- for full checks, allowing
exceptions, goto, inlining, and prefer classic strings. }
{$longstrings off}            { to be sure, use AnsiString or pchar explicitly }
{$booleval on}                { standard pascal expects this }

type
   integer = int64;            { by default ints are 64 bits this decade}

{ advent of code 2023 day 02 -- cube conundrum
  drawing colored cubes from a bag, how many draws can be made.

  input format:

  game #: sample (;sample)*

  sample is # color (, # color)*
  color is red | blue | green
  # is an integer }


const
   MAXSAMPLES = 10;


type

   tSample = record                { record of cubes drawn }
                red   : integer;   { counted by color      }
                green : integer;
                blue  : integer;
             end;

{ a processed line from the input dataset. }

   tPayload = record
                 original : string;    { see input desc above }
                 gameno   : integer;   { unique id }
                 count    : integer;   { of samples }
                 samples  : array [1 .. MAXSAMPLES] of tSample;
              end;

{ the input dataset for the days problem, one per line of input. the
  original input line and its parsed value(s). blank lines are
  maintained for control break purposes. }

   pData = ^tData;
   tData = record
              fwd,              { a singly linked list would be }
              bwd : pData;      { enough, but you never know }
              num : integer;    { line number in source }
              rec : string;     { original input }
              pld : tPayload;   { what does it mean stimpy? }
           end;


{$I aocinput.p }
{$I max.p }
{$I str2int.p }


{ parse the input record (see description above) into a payload for part one. }

procedure parseone(var pld : tPayload);

var
   i, p : integer;

begin
   with pld do begin
      for i := 1 to MAXSAMPLES do
         with samples[i] do begin
            red := 0;
            blue := 0;
            green := 0
         end; {with samples[i] }
      p := 6;                  { position after 'game ' }
      gameno := str2int(original, p);
      count := 0;
      while (p <= length(original)) and (count <= MAXSAMPLES) and (original[p] in [';', ':']) do begin
         count := count + 1;    { next sample slot }
         repeat
            p := p + 2;          { skip [:;]\s }
            i := str2int(original, p); { get number cubes }
            p := p + 1;          { skip to color }
            case original[p] of
              'r' : begin
                 samples[count].red := i;
                 p := p + 3; { skip ' red' }
              end; { 'r' }
              'g' : begin
                 samples[count].green := i;
                 p := p + 5; { skip 'green' }
              end;
              'b' : begin
                 samples[count].blue := i;
                 p := p + 4; {skip 'blue' }
              end;
            end; { case }
         until original[p] <> ',';
      end { while pos/count/check delimiter }
   end
end;


{ load and parse the problem dataset }

function load : pData;

var
   first, prior, curr : pData;
   i                  : integer;
   s                  : string;

begin
   aocrewind;
   first := nil;
   prior := nil;
   curr := nil;
   i := 0;
   while not aoceof do begin
      i := i + 1;
      s := aocread;
      new(curr);
      curr^.fwd := nil;
      curr^.bwd := prior;
      curr^.num := i;
      curr^.rec := s;
      curr^.pld.original := s;
      parseone(curr^.pld);
      if first = nil then
         first := curr
      else
         prior^.fwd := curr;
      prior := curr
   end; { while not eof }
   load := first;
end;


{ for part one, the game draws of input are checked to see if they
  would be possible from a bag loaded with only 12 red cubes, 13
  green cubes, and 14 blue cubes. The answer for part one is the
  sum of the game ids of possible games. }

function partone(pd : pData) : integer;

var
   bag  : tSample;
   i, j : integer;

begin
   partone := 0;
   bag.red := 12;    { sample for part one test }
   bag.green := 13;
   bag.blue := 14;
   i := 0;
   while pd <> nil do
      with pd^.pld do begin
         i := i + 1;
         partone := partone + gameno;   { assume this passes }
         for j := 1 to count do
            if (bag.red < samples[j].red) or
               (bag.green < samples[j].green) or
               (bag.blue < samples[j].blue) then begin
                  partone := partone - gameno; { assumption wrong }
                  break;
               end;
         pd := pd^.fwd
      end; { with pd^ }
end;


{ for part two, determine the least number of cubes can satisfy a
  game's draws. for this set, multiply the number of cubes together,
  and return the sum. }

function parttwo(pd : pData) : integer;

var
   bag         : tSample;
   power, i, j : integer;

begin
   parttwo := 0;
   i := 0;
   while pd <> nil do
      with pd^.pld do begin
         i := i + 1;
         bag.red := 0;       { initialize for min }
         bag.green := 0;
         bag.blue := 0;
         for j := 1 to count do begin
            bag.red := max(bag.red, samples[j].red);
            bag.green := max(bag.green, samples[j].green);
            bag.blue := max(bag.blue, samples[j].blue);
         end;
         power := bag.red * bag.green * bag.blue;
         parttwo := parttwo + power;
         pd := pd^.fwd
      end; { with pd^ }
end;


{ mainline }

var
   i       : integer;
   dataset : pData;

begin
   aocopen;
   dataset := load;
   i := partone(dataset);
   writeln('part one: ', i);
   i := parttwo(dataset);
   writeln('part two: ', i);
   aocclose;
end.

File: 2023/02/solution.f90
! solution.f90 -- 2023 day 02 -- cube conundrum -- troy brumley blametroi@gmail.com

! advent of code 2023 day 02 -- cube conundrum
!
!  drawing colored cubes from a bag, how many draws can be made?
!
! input format:
!
! game #: sample (;sample)*
!
!   sample is # color (, # color)*
!   color is red | blue | green
!   # is an integer
!
! The Elf would first like to know which games would have been possible if the bag
! contained only 12 red cubes, 13 green cubes, and 14 blue cubes?

program solution

   use iso_fortran_env, only: int64
   use aocinput, only: open_aoc_input, read_aoc_input, close_aoc_input, max_aoc_reclen, rewind_aoc_input

   implicit none

   integer, parameter            :: AOCIN = 10
   integer(kind=int64)           :: part_one, part_two  ! results
   character(len=max_aoc_reclen) :: rec

   ! domain data and constants

   integer, parameter  :: MAX_SAMPLES = 10
   integer, parameter  :: MAX_GAMES = 100

   character(len=5), parameter :: color_names(3) = ["red  ", "green", "blue "]

   type sample_t
      integer          :: red
      integer          :: green
      integer          :: blue
   end type sample_t

   type game_t
      integer         :: num
      integer         :: count
      type(sample_t)  :: sample(1:MAX_SAMPLES)
   end type game_t

   type(game_t)       :: game(1:MAX_GAMES)
   integer            :: numgames

   ! initialize
   rec = ""
   call open_aoc_input(AOCIN)
   part_one = 0; part_two = 0
   numgames = 0

   ! load
   do while (read_aoc_input(AOCIN, rec))
      if (len_trim(rec) < 1) cycle
      numgames = numgames + 1
      game(numgames) = parse()
   end do

   ! and find our answers
   part_one = dopartone()
   part_two = doparttwo()

   ! report and close
   print *
   print *, part_one, "part one"
   print *, part_two, "part two"
   print *

   call close_aoc_input(AOCIN)

contains

   ! simple predicates and converters

   function is_digit(c) result(res)
      implicit none
      character(len=1), intent(in) :: c
      logical                      :: res
      res = c <= "9" .and. c >= "0"
   end function is_digit

   function char_int(c) result(res)
      implicit none
      character(len=1), intent(in) :: c
      integer                      :: res
      res = -1
      if (is_digit(c)) res = ichar(c) - ichar("0")
   end function char_int

   ! parse an input record into a game structure. assumes input is "rec" and
   ! returns a populated game_t.
   !
   ! game #: sample (;sample)*
   !
   !   sample is # color (, # color)*
   !   color is red | blue | green
   !   # is an integer

   function parse() result(res)
      implicit none
      character(len=:), allocatable :: s
      type(game_t)                  :: res
      integer                       :: i, pos
      type(sample_t)                :: zero_sample

      ! an empty template
      zero_sample % red = 0
      zero_sample % green = 0
      zero_sample % blue = 0

      ! some work on the record to ease parsing. replace all the trailing
      ! blanks with "semicolon<blank>". you no longer need to trim when
      ! checking length.
      allocate (character(len=len_trim(rec) + 2) :: s)
      s = trim(rec)//"; "

      ! extract game number, reacord header is "game #: "

      res % num = 0
      pos = 6 ! "game "
      gameno: do while (pos < len(s) .and. is_digit(s(pos:pos)))
         res % num = res % num * 10 + char_int(s(pos:pos))
         pos = pos + 1
      end do gameno
      ! pos is now at ": "

      ! gather each sample, the parser is very trusting and will fall all to
      ! pieces if you feed it the wrong file.

      res % count = 0
      pos = pos + 2 ! first digit of first sample color

      samples: do while (pos < len(s))
         res % count = res % count + 1
         res % sample(res % count) = zero_sample

         colors: do

            ! find start of color count
            seekdigit: do while (.not. is_digit(s(pos:pos)))
               pos = pos + 1
            end do seekdigit
            ! pos is at first digit

            ! convert
            i = 0
            accumint: do while (is_digit(s(pos:pos)))
               i = i * 10 + char_int(s(pos:pos))
               pos = pos + 1
            end do accumint
            pos = pos + 1
            ! pos is now at the first letter of the color name

            ! as colors are uniquely identified by the first letter, we're not
            ! checking the whole string.

            if (s(pos:pos) == "r") then ! red
               res % sample(res % count) % red = i
               pos = pos + 3
            else if (s(pos:pos) == "g") then ! green
               res % sample(res % count) % green = i
               pos = pos + 5
            else if (s(pos:pos) == "b") then ! blue
               res % sample(res % count) % blue = i
               pos = pos + 4
            else ! what, you thought i'd just assume blue above?
               print *, "*** error *** expected color at position", pos, "not found '", trim(rec), "'"
               stop 1
            end if
            ! pos is now at first character after color name

            ! pos is now on either a comma, meaing another color is in this sample, or a
            ! semicolon, meaning that was the of this sample
            if (s(pos:pos + 1) == "; ") then
               pos = pos + 2
               cycle samples
            else if (s(pos:pos + 1) == ", ") then
               pos = pos + 2
               cycle colors
            end if

            print *, "*** error *** expected comma or semicolon at position", pos, "not found '", trim(rec), "'"
            stop 1
         end do colors
      end do samples
   end function parse

   ! for part one the game draws of input are checked to see if the would be
   ! possible from a bag loaded with only 12 red cubes, 13 green cubes, and 14
   ! blue cubes. the answer for part one is the sum of game numbers of the
   ! possible games.

   function dopartone() result(res)
      implicit none
      integer :: res

      type(sample_t) :: bag
      integer :: i, j

      ! available cubes as defined in the problem statement
      bag % red = 12
      bag % green = 13
      bag % blue = 14

      res = 0
      i = 0
      do i = 1, NUMGAMES
         !print *, "game", i
         ! assume this passes
         res = res + game(i) % num
         samples: do j = 1, game(i) % count
            !print *, "sample", j, game(i) % sample(j) % red, game(i) % sample(j) % green, game(i) % sample(j) % blue
            if ((bag % red < game(i) % sample(j) % red) .or. &
                (bag % green < game(i) % sample(j) % green) .or. &
                (bag % blue < game(i) % sample(j) % blue)) then
               !print *, "bzzt"
               ! the assumption has proved wrong
               res = res - game(i) % num
               exit samples
            end if
         end do samples
      end do
   end function dopartone

! for part two we must find the fewest number of cubes can satisfy a
! game's draws. the answer for part two is sum of the products of the cubes
! for each game.

   function doparttwo() result(res)
      implicit none
      integer :: res
      type(sample_t) :: bag
      integer :: power, i, j

      res = 0
      do i = 1, numgames

         bag % red = 0 ! initialize to empty
         bag % green = 0
         bag % blue = 0

         do j = 1, game(i) % count
            bag % red = max(bag % red, game(i) % sample(j) % red)
            bag % green = max(bag % green, game(i) % sample(j) % green)
            bag % blue = max(bag % blue, game(i) % sample(j) % blue)
         end do

         power = bag % red * bag % green * bag % blue
         res = res + power
      end do
   end function doparttwo

end program solution
